subsys/debug/tracing/cpu_stats.c

```diff
@@ -0,0 +1,214 @@
+/*
+ * Copyright (c) 2018 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <tracing_cpu_stats.h>
+#include <misc/printk.h>
+
+enum cpu_state {
+	CPU_STATE_IDLE,
+	CPU_STATE_NON_IDLE,
+	CPU_STATE_SCHEDULER
+};
+
+static enum cpu_state last_cpu_state = CPU_STATE_SCHEDULER;
+static enum cpu_state cpu_state_before_interrupts;
+
+static u32_t last_time;
+static u64_t idle_cnt;
+static u64_t non_idle_cnt;
+static u64_t sched_cnt;
+static int nested_interrupts;
+static struct k_thread *current_thread;
+
+#ifndef CONFIG_SMP
+extern k_tid_t const _idle_thread;
+#endif
+
+static inline int is_idle_thread(struct k_thread *thread)
```

Comment:
There is already `_is_idle()` in sched.c.
IMHO we should re-use the code instead of copying it around.

Reply:
this function is static inline, so we cannot simply use it in our files;
it is done like this for sysview profiler hooks

Reply:
if we can limit the number of calls at this level it would be great, so I recommend to leave it as is.

Reply:
Could we move the `_is_idle()` to some header and make it public for other kernel parts?
@andrewboie: What do you think?

Reply:
i already asked @andyross to make this a function call. We can clean this up once this function is available.

Reply:
@pizi-nordic let's fix it later when API is changed. We need this change.